File: frontend/src/services/budgetService.ts
import axiosClient from './axiosClient';

export interface Budget {
  id: string;
  category: string;
  amountLimit: number;
  spent: number;
  month: string;
  createdAt: string;
}

export interface FinancialGoal {
  id: string;
  name: string;
  targetAmount: number;
  currentAmount: number;
  percentageProgress: number;
  targetDate: string;
  status: string;
  estimatedCompletionText: string;
  createdAt: string;
}

export const budgetService = {
  getBudgets: async (month?: string): Promise<Budget[]> => {
    const response = await axiosClient.get('/api/budget', { params: { month } });
    return response.data.data;
  },

  createBudget: async (data: { category: string; amountLimit: number; month: string }): Promise<Budget> => {
    const response = await axiosClient.post('/api/budget', data);
    return response.data.data;
  },

  updateBudget: async (id: string, data: { amountLimit: number; category: string; month: string }): Promise<Budget> => {
    const response = await axiosClient.put(`/api/budget/${id}`, data);
    return response.data.data;
  },

  deleteBudget: async (id: string): Promise<void> => {
    await axiosClient.delete(`/api/budget/${id}`);
  },

  getGoals: async (): Promise<FinancialGoal[]> => {
    const response = await axiosClient.get('/api/budget/goal');
    return response.data.data;
  },

  createGoal: async (data: { name: string; targetAmount: number; currentAmount: number; targetDate: string }): Promise<FinancialGoal> => {
    const response = await axiosClient.post('/api/budget/goal', data);
    return response.data.data;
  },

  updateGoal: async (id: string, data: { name: string; targetAmount: number; currentAmount: number; targetDate: string }): Promise<FinancialGoal> => {
    const response = await axiosClient.put(`/api/budget/goal/${id}`, data);
    return response.data.data;
  },

  deleteGoal: async (id: string): Promise<void> => {
    await axiosClient.delete(`/api/budget/goal/${id}`);
  }
};
export default budgetService;

File: backend/src/main/java/com/apexpay/service/BudgetService.java
package com.apexpay.service;

import com.apexpay.dto.BudgetRequest;
import com.apexpay.dto.BudgetResponse;
import com.apexpay.dto.FinancialGoalRequest;
import com.apexpay.dto.FinancialGoalResponse;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

public interface BudgetService {
    BudgetResponse createBudget(UUID userId, BudgetRequest request);
    BudgetResponse updateBudget(UUID id, UUID userId, BudgetRequest request);
    List<BudgetResponse> getBudgets(UUID userId, String month);
    void deleteBudget(UUID id, UUID userId);
    void checkAndUpdateBudgetsOnTransaction(UUID userId, String category, BigDecimal amount, String month);
    
    FinancialGoalResponse createGoal(UUID userId, FinancialGoalRequest request);
    FinancialGoalResponse updateGoal(UUID id, UUID userId, FinancialGoalRequest request);
    List<FinancialGoalResponse> getGoals(UUID userId);
    void deleteGoal(UUID id, UUID userId);
}
